tests/unit/test.watcher.js

```diff
@@ -207,3 +207,48 @@ describe('watcher', () => {
   });
 
 });
+
+describe('--watch-ignored is passed in', () => {
+  it('does not change if ignored file is touched', () =>
+    withTempDir(async (tmpDir) => {
+      const debounceTime = 10;
+      const onChange = sinon.spy();
+      const tmpPath = tmpDir.path();
+      const files = ['foo.txt', 'bar.txt', 'foobar.txt'].map(
+        (filePath) => path.join(tmpPath, filePath)
+      );
+
+      const watcher = onSourceChange({
+        sourceDir: tmpPath,
+        artifactsDir: path.join(tmpPath, 'web-ext-artifacts'),
+        onChange,
+        watchIgnored: ['foo.txt'].map(
+          (filePath) => path.join(tmpPath, filePath)
+        ),
+        shouldWatchFile: () => true,
+        debounceTime,
+      });
+
+      async function waitDebounce() {
+        await new Promise((resolve) => setTimeout(resolve, debounceTime * 2));
+      }
+
+      // Verify foo.txt is being ignored.
+      await fs.writeFile(files[0], '<content>');
+      sinon.assert.notCalled(onChange);
+
+      // Verify that the other two files are not be ignored.
+      await fs.writeFile(files[1], '<content>');
+      await waitDebounce();
+      sinon.assert.calledOnce(onChange);
```

Comment:
> @rpl, I am not sure why onChange function is never being called even if we write to a file that is not being ignored.
> After doing some debugging, I found that `watcher.on('change')` event is never being invoked
> 
> https://github.com/mozilla/web-ext/blob/afd9592579ceb9606f2defe9afa9446c3f8227c5/src/watcher.js#L50-L52
> 
> **Doubt**:
> 
>     1. What difference does it make when we use `watchChange` helper function instead of using `onSourceChange ` directly?   
>, why does it do intial scan when we use `watchChange`?

Totally reasonable doubts, let me try to provide some more details:

The watchChange helper does currently enforce a certain order of the steps, and the file is written before the watcher is going to start monitoring the directory for changes and then updated, that means that the initialScan will happen when the directory being watched already contains the file.

By using the onSourceChange directly we are ensure that we are in control of the steps (without having to further increase the complexity of that helper) and that no files are in the directory being watched and we write them later to cover a couple of expected behaviors around the test case scenario.

Let me know if the details above are answering your questions or if there are other points unclear.

> 
> 
> Thank you!...
> 
> **Update**: I see that tests on travis-ci are passing successfully but on my machine `--watch-ignored` test is failing. Here's a screenshot of that:
> **On linux, tests are passing successfully**

yeah, watching files is one of those things that it is not unlikely to behave differently based on the os (and also the particular hardware)

I'm also afraid that, giving that it does look like a "race condition" kind of issue, it may also start to fail intermittently on the CI over an higher number of builds.

I think that it would be a good idea to look into an alternative strategy that may not be actually subjected to race-related failures that easily.

One possible other strategy could be to use a secondary watcher instance that does not have any watchIgnored parameter, so we can wait for the change event to have been emitted on that additional watcher instance.

Interesting enough, when I tried that locally I did noticethat I had to add a short delay (using `await waitDebounce();`) before writing the file, otherwise none of the two watcher was actually emitting the 'change' event we expect, this match what you did notice by debugging the failure you get on macos.
Nevertheless using the secondary watcher may still be a good idea to make the test as stable as possible (and avoid intermittent failures).

Would you mind to give that a try also on your machine? so that we can double-check that this strategy does make the test to pass consistently also on macos as well as on Linux.

Here is the diff of the changes I did apply locally:

```diff --git a/tests/unit/test.watcher.js b/tests/unit/test.watcher.js
index 03df40b..8f4c833 100644
--- a/tests/unit/test.watcher.js
+++ b/tests/unit/test.watcher.js
@@ -6,6 +6,7 @@ import {it, describe} from 'mocha';
 import {fs} from 'mz';
 import sinon from 'sinon';
 import {assert} from 'chai';
+import Watchpack from 'watchpack';
 
 import {default as onSourceChange, proxyFileChanges} from '../../src/watcher';
 import {withTempDir} from '../../src/util/temp-dir';
@@ -229,24 +230,33 @@ describe('--watch-ignored is passed in', () => {
         debounceTime,
       });
 
+      const watchAll = new Watchpack();
+      watchAll.watch(files, [], 0);
+
       async function waitDebounce() {
         await new Promise((resolve) => setTimeout(resolve, debounceTime * 2));
       }
 
+      async function assertOnChange(filePath, expectedCallCount) {
+        const promiseOnChanged = new Promise((resolve) => watchAll.once(
+          'change',
+          (f) => resolve(f))
+        );
+        await waitDebounce();
+        await fs.writeFile(filePath, '<content>');
+        assert.equal(filePath, await promiseOnChanged);
+        sinon.assert.callCount(onChange, expectedCallCount);
+      }
+
       // Verify foo.txt is being ignored.
-      await fs.writeFile(files[0], '<content>');
-      sinon.assert.notCalled(onChange);
+      await assertOnChange(files[0], 0);
 
       // Verify that the other two files are not be ignored.
-      await fs.writeFile(files[1], '<content>');
-      await waitDebounce();
-      sinon.assert.calledOnce(onChange);
-
-      await fs.writeFile(files[2], '<content>');
-      await waitDebounce();
-      sinon.assert.calledTwice(onChange);
+      await assertOnChange(files[1], 1);
+      await assertOnChange(files[2], 2);
 
       watcher.close();
+      watchAll.close();
       // Leave watcher.close some time to complete its cleanup before withTempDir will remove the
       // test directory.
       await waitDebounce();
```

Let me know how it goes, this test case turned out to be more interesting than expected ;-)